### PR TITLE
Microservice Product Search for searching in Elasticsearch

### PR DESCRIPTION
**Pull request by Miroslav-Stopka**
| Q             | A
| ------------- | ---
|Description, reason for the PR| Experimental microservice for product search via Elasticsearch, integrated with the Shopsys Framework.
|New feature| Yes
|BC breaks| Yes
|Fixes issues| ...
|Standards and tests pass| Yes
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: docs/installation/installation-using-docker-application-setup.md [hunk 1]
```diff
@@ -9,21 +9,26 @@ Clone the repository inside a separate directory and configure the container in
...
         links:
-            - postgres
+            - elasticsearch
         depends_on:
-            - postgres
+            - elasticsearch
 ```
-*Note: If you have edited the docker-compose.yml, you should re-run the docker-compose command --force-recreate -d
-*Note: There can be differences in the configuration on different OS. It should be similar to the `php-fpm`'s configuration.*
+Edit the path `../microservices/product-search` in such a way as to direct it to the directory with the cloned microservice.
```
Comment (TomasLudvik) [inline docs/installation/installation-using-docker-application-setup.md]: I do not understand this sentence

File: microservices/product-search/bin/console [hunk 2]
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env php
```
Comment (TomasLudvik) [inline microservices/product-search/bin/console]: It would be nice if Symfony skeleton could be presented in a separate commit, so we know exactly what has changed by microservice

File: docs/installation/installation-using-docker-application-setup.md [hunk 3]
```diff
@@ -3,15 +3,58 @@
 This guide expects that you have already set up your Docker environment.
 If you haven't already done that check the [Installation Using Docker](installation-using-docker.md).
 
-## 1. Setup the application
-Now that the Docker environment is prepared we can setup the application itself.
+## 1. Setup the microservice for product search
+Installing of Shopsys Framework from a project-base requires manual cloning of the [repository of the microservice](https://github.com/shopsys/microservice-product-search).
+Clone the repository inside a separate directory and configure the container in your `docker-compose.yml`:
+```yaml
```
Comment (TomasLudvik) [inline docs/installation/installation-using-docker-application-setup.md]: Why is this part of documentation when it is already set in .dist docker-compose files? I would expect this to be maybe part of the UPGRADE.md so people know how to upgrade.

File: docs/installation/native-installation.md [hunk 4]
```diff
@@ -52,6 +53,25 @@ monolog:
...
+It will act as a fully independent unit with a separate web server and repository.
+
+Just clone the repository into a separate directory, install all its dependencies, configure connection to the Postgres by setting up the PG_CONNECTION_STRING environment variable (or the [.env file](http://symfony.com/doc/current/components/dotenv.html)) and run the server.
+In the current version of this experimental microservice, the PHP's built-in Web Server is used.
+
+```
+git clone https://github.com/shopsys/microservice-product-search.git
+cd microservice-product-search
+composer install
+# open ".env" and set up connection to Postgres in "PG_CONNECTION_STRING"
```
Comment (TomasLudvik) [inline docs/installation/native-installation.md]: this comment is weird, it says:
php bin/console server:run 127.0.0.1:8001 is doing open ".env" and set up connection to Postgres in "PG_CONNECTION_STRING" 

This needs to be reworked

File: microservices/product-search/.env.dist [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# This file is a "template" of which env vars need to be defined for your application
+# Copy this file to .env file for development, create environment variables when deploying to production
+# https://symfony.com/doc/current/best_practices/configuration.html#infrastructure-related-configuration
+
+###> symfony/framework-bundle ###
+APP_ENV=dev
+APP_SECRET=bef043dff42c07c95949d8bfddf1081f
+#TRUSTED_PROXIES=127.0.0.1,127.0.0.2
```
[the remark this answers is not in the document]
Comment (PetrHeinz) [reply microservices/product-search/.env.dist]: It's just an example added here during the symfony/skeleton installation. You can see the `#` at the beginning of the line, it's just a comment.

File: docs/installation/installation-using-docker-application-setup.md [hunk 6]
```diff
@@ -3,15 +3,58 @@
 This guide expects that you have already set up your Docker environment.
 If you haven't already done that check the [Installation Using Docker](installation-using-docker.md).
 
-## 1. Setup the application
-Now that the Docker environment is prepared we can setup the application itself.
+## 1. Setup the microservice for product search
```
[the remark this answers is not in the document]
Comment (Miroslav-Stopka) [reply docs/installation/installation-using-docker-application-setup.md]: What is ugly? You mean these ultra simple instructions how to use this microservice? Please, can you be more specific?

And Why 1.1 is better than 1? 

Of course, you are right. The correct way is microservice as a image. But, right now, we have zero experiences with this approach. That's why we decided to choose this way. Iterative development is about solving small problems in every iteration. In this iteration of microservice we have tried to solve the problems of "creation of the microservice" and "environment of the microservice". Thanks to this approach, after only 3 weeks we have absolutely fully functional microservice with PR for merging it with elastic.  After this iteration will be completed, then we will start to solve the problems "microservice as image" and "how the developer of the project will modify the microservice that we provide as an image".

File: docs/installation/installation-using-docker-application-setup.md (hunk 3 above)
Comment (Miroslav-Stopka) [reply docs/installation/installation-using-docker-application-setup.md]: I decided to let the developers know what exactly changed in the docker-compose. Do you think that this should be removed? If yes, I will remove it.

File: docs/installation/installation-using-docker-application-setup.md [hunk 7]
```diff
@@ -3,15 +3,58 @@
...
+```
+composer install
+```
+### 1.3. Run the server
+Since the microservice acts as a fully independent unit, it needs a separate web server.
+In the current version of this experimental microservice, the PHP's built-in Web Server is used.
+
+You can start the server by executing the command:
+```
+php bin/console server:run *:8000
```
[the remark this answers is not in the document]
Comment (Miroslav-Stopka) [reply docs/installation/installation-using-docker-application-setup.md]: We have user story for this problem. A solution of the problem related to Servers for Microservices is not an easy one to find. Some recommendations speak about one container with microservice and server. Another speaks about separated containers for every microservice and every server for microservice. This should not be decided by one person and it needs to be analysed and explored more deeply. 

Basically, this fully functional microservice will be iteratively improved and change of the servers definitely one of the improvements .

File: docs/installation/native-installation.md (hunk 4 above)
Comment (Miroslav-Stopka) [reply docs/installation/native-installation.md]: Im not sure, I dont see "php bin/console server:run 127.0.0.1:8001 is doing open"? Can you be more specific? Thank you

File: microservices/product-search/README.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
+# Microservice Product Search
+
+Microservice responsible for product search on [Shopsys Framework](https://www.shopsys-framework.com).
```
[the remark this answers is not in the document]
Comment (Miroslav-Stopka) [reply microservices/product-search/README.md]: Can you come with some reason, please? This sentence says "Microservisa zodpovědná za hledání produktu".

File: microservices/product-search/README.md [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+{"productIds":[2,1,3]}
+```
+
+## Logging
+Microservice Product Search is a Symfony based application that uses [Monolog](https://github.com/Seldaek/monolog) with [symfony/monolog-bundle](https://github.com/symfony/monolog-bundle) as a logging tool.
+By default, the logs are routed into a named pipe `/tmp/log-pipe` (the same way as [Logging in the main application](https://github.com/shopsys/shopsys/blob/master/docs/introduction/logging.md) works).
+
+To see the logs simply run: 
+```
+docker-compose logs
```
[the remark this answers is not in the document]
Comment (Miroslav-Stopka) [reply microservices/product-search/README.md]: docker logs will show you all logs of all containers. It means even the logs from php-fpm container, which can provide some information about error in communication with microservice. That's why I think it's better for users to use this command. If they want to display only pieces of information from a microservice container, they can grep 'microservice'

File: microservices/product-search/config/services.yaml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+        exclude: '../src/{DependencyInjection,Entity,Migrations,Tests,Kernel.php}'
+
+    # controllers are imported separately to make sure services can be injected
+    # as action arguments even if you don't extend any base controller class
+    Shopsys\MicroserviceProductSearch\Controller\:
+        resource: '../src/Controller'
+        tags: ['controller.service_arguments']
+
+    # add more service definitions when explicit configuration is needed
+    # please note that last definitions always *replace* previous ones
```
Comment (Miroslav-Stopka) [inline microservices/product-search/config/services.yaml]: This comments will be removed :)

File: microservices/product-search/config/routes.yaml [hunk 11]
```diff
@@ -1,3 +1,5 @@
-#index:
-#    path: /
-#    controller: App\Controller\DefaultController::index
+search_product_ids:
+  path: /{domainId}/product-ids
```
[the remark this answers is not in the document]
Comment (Miroslav-Stopka) [reply microservices/product-search/config/routes.yaml]: Route will be changed to search-product-ids

File: build.xml [hunk 12]
```diff
@@ -33,8 +33,8 @@
         <phingcall target="composer-check" />
     </target>
 
-    <target name="standards" depends="phplint,ecs,phpstan,twig-lint,yaml-lint,eslint-check,standards-packages" description="Checks coding standards in both project-base and packages."/>
-    <target name="standards-diff" depends="phplint-diff,ecs-diff,phpstan,twig-lint-diff,yaml-lint,eslint-check-diff,standards-packages" description="Checks coding standards on changed files in project-base and all files in packages."/>
+    <target name="standards" depends="phplint,ecs,phpstan,twig-lint,yaml-lint,eslint-check,standards-packages" description="Checks coding standards in the project-base, packages and microservices."/>
```
[the remark this answers is not in the document]
Comment (Miroslav-Stopka) [reply build.xml]: So the question is, why microservices are ignored in phpstan right now.

File: docs/installation/installation-using-docker-application-setup.md (hunk 3 above)
Comment (Miroslav-Stopka) [reply docs/installation/installation-using-docker-application-setup.md]: I will remove it.

File: docs/installation/native-installation.md (hunk 4 above)
Comment (Miroslav-Stopka) [reply docs/installation/native-installation.md]: I see now, these commands will be splitted

Comment (TomasLudvik) [reply docs/installation/native-installation.md]: We have discussed this

File: microservices/product-search/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (Miroslav-Stopka) [reply microservices/product-search/README.md]: docker-compose logs | grep 'microservice'

File: microservices/product-search/src/Repository/ProductSearchRepository.php [hunk 13]
```diff
@@ -2,74 +2,78 @@
...
+     * @param \Elasticsearch\Client $client
      */
-    private $pgConnectionString;
+    public function __construct(Client $client)
+    {
+        $this->client = $client;
+    }
 
-    public function __construct(TsqueryFactory $tsqueryFactory, string $pgConnectionString)
+    public function getFoundProductIds(int $domainId, string $searchText): array
```
[the remark this answers is not in the document]
Comment (Miroslav-Stopka) [reply microservices/product-search/src/Repository/ProductSearchRepository.php]: getProductIdsBySearchText

File: microservices/product-search/src/Repository/ProductSearchRepository.php [hunk 14]
```diff
@@ -2,74 +2,78 @@
...
-        $parameters = [
-            '%' . DatabaseSearching::getLikeSearchString($searchText) . ' %',
-            '%' . DatabaseSearching::getLikeSearchString($searchText),
-            $this->tsqueryFactory->getTsqueryWithAndConditions($searchText),
-            $this->tsqueryFactory->getTsqueryWithOrConditions($searchText),
-            DatabaseSearching::getFullTextLikeSearchString($searchText),
-            $this->tsqueryFactory->getTsqueryWithAndConditionsAndPrefixMatchForLastWord($searchText),
-            $this->tsqueryFactory->getTsqueryWithOrConditionsAndPrefixMatchForLastWord($searchText),
-            $domainId,
+        return [
```
[the remark this answers is not in the document]
Comment (Miroslav-Stopka) [reply microservices/product-search/src/Repository/ProductSearchRepository.php]: "@See https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-body.html" will be added into annotation

File: docs/installation/installation-using-docker-application-setup.md (hunk 1 above)
Comment (Miroslav-Stopka) [reply docs/installation/installation-using-docker-application-setup.md]: ... so it leads to the directory ...

Comment (TomasLudvik) [reply docs/installation/installation-using-docker-application-setup.md]: Edit the path `../microservices/product-search` in such way so it leads to the directory with the cloned microservice.

File: microservices/product-search/src/Kernel.php [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    protected function configureContainer(ContainerBuilder $container, LoaderInterface $loader)
+    {
+        $container->addResource(new FileResource($this->getProjectDir().'/config/bundles.php'));
+        // Feel free to remove the "container.autowiring.strict_mode" parameter
+        // if you are using symfony/dependency-injection 4.0+ as it's the default behavior
+        $container->setParameter('container.autowiring.strict_mode', true);
+        $container->setParameter('container.dumper.inline_class_loader', true);
+        $confDir = $this->getProjectDir().'/config';
+
+        $loader->load($confDir.'/{packages}/*'.self::CONFIG_EXTS, 'glob');
```
Comment (TomasLudvik) [inline microservices/product-search/src/Kernel.php]: I thought that this was already repaired (spaces around dots

File: homebrew/CODE_OF_CONDUCT.md [hunk 16]
```diff
@@ -0,0 +1,25 @@
+# Code of Conduct
```
Comment (TomasLudvik) [inline homebrew/CODE_OF_CONDUCT.md]: This should not be here

File: homebrew/CHANGELOG.md [hunk 17]
```diff
@@ -0,0 +1 @@
+See Homebrew's [releases on GitHub](https://github.com/Homebrew/brew/releases) for the changelog.
```
Comment (TomasLudvik) [inline homebrew/CHANGELOG.md]: This should not be here

File: homebrew/bin/brew [hunk 18]
```diff
@@ -0,0 +1,82 @@
+#!/bin/bash
```
Comment (TomasLudvik) [inline homebrew/bin/brew]: This should not be here

File: build.xml (hunk 12 above)
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply build.xml]: and the answer is?

File: microservices/product-search/README.md [hunk 19]
```diff
@@ -53,7 +53,7 @@ Some attributes of a product can be different across the domains.
 
 Example:
 ```php
-$this->microserviceProductSearchClient->get('product-ids', [
+$this->microserviceProductSearchClient->get('search-product-ids', [
```
Comment (TomasLudvik) [inline microservices/product-search/README.md]: could you make this change in an appropriate commit? (I think the first or second one it is)

**Comment by TomasLudvik** [comment]
Oh, one more thing, pls rebase it on master

**Comment by Miroslav-Stopka** [comment]
Rebase on master - done :)

File: microservices/product-search/README.md (hunk 19 above)
Comment (Miroslav-Stopka) [reply microservices/product-search/README.md]: Done :)

File: microservices/product-search/src/Kernel.php (hunk 15 above)
Comment (Miroslav-Stopka) [reply microservices/product-search/src/Kernel.php]: Look at Changed files, there you will see the space around dots :)

File: homebrew/bin/brew (hunk 18 above)
Comment (Miroslav-Stopka) [reply homebrew/bin/brew]: Done

File: homebrew/CHANGELOG.md (hunk 17 above)
Comment (Miroslav-Stopka) [reply homebrew/CHANGELOG.md]: Done

File: homebrew/CODE_OF_CONDUCT.md (hunk 16 above)
Comment (Miroslav-Stopka) [reply homebrew/CODE_OF_CONDUCT.md]: Done

File: docs/installation/installation-using-docker-application-setup.md [hunk 20]
```diff
@@ -3,15 +3,49 @@
...
 If you haven't already done that check the [Installation Using Docker](installation-using-docker.md).
 
-## 1. Setup the application
-Now that the Docker environment is prepared we can setup the application itself.
+## 1. Setup the microservice for product search
+Installing of Shopsys Framework from a project-base requires manual cloning of the [repository of the microservice](https://github.com/shopsys/microservice-product-search).
+
+Clone the repository inside a separate directory and configure the new container `microservice-product-search` in your `docker-compose.yml` according to the updated templates in [project-base/docker/conf](https://github.com/shopsys/project-base/tree/master/docker/conf).
+
+Edit the path `../microservices/product-search` of the new container in such a way so it leads to the directory with the cloned microservice.
```
Comment (LukasHeinz) [inline docs/installation/installation-using-docker-application-setup.md]: I suggest to add a note somewhere that this issue will be addressed soon.

File: microservices/product-search/README.md [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+## How it works
+Microservice acts as a fully independent unit.
+It uses its own separate server, separate logic and it uses own vendor directory for Composer dependencies.
+All of these parts are located in a separate Docker container.
+This container is configured in [docker-compose.yml](https://github.com/shopsys/shopsys/tree/master/docker/conf) under the name `microservice-product-search`
+
+As the base of the microservice, we used micro-kernel which is made up of [Symfony 4](https://symfony.com/4).
+This leanest Symfony version is highly optimized and it is suitable for this type of project.
+
+For the search process is used a super fast no-SQL database Elasticsearch. For more information about the Elasticsearch on Shopsys Framework, see [Product search via Elasticsearch]((https://github.com/shopsys/shopsys/blob/master/docs/introduction/product-search-via-elasticsearch.md).)
```
Comment (LukasHeinz) [inline microservices/product-search/README.md]: There seems to be misplaced parenthesis in the end -> the link doesn't work

Comment (Miroslav-Stopka) [reply microservices/product-search/README.md]: This status is correct. README.md is in a different repo. So you can not use a relative path. This link will be functional after merging into master. You can use link into the current branch, but after merge into master and after removal of this old branch, your link will be broken.
